compiler/rustc_hir/src/hir.rs

```diff
@@ -526,9 +525,17 @@ pub struct GenericParam<'hir> {
 impl GenericParam<'hir> {
     pub fn bounds_span(&self) -> Option<Span> {
         self.bounds.iter().fold(None, |span, bound| {
-            let span = span.map(|s| s.to(bound.span())).unwrap_or_else(|| bound.span());
-
-            Some(span)
+            if let ExpnKind::Macro(MacroKind::Derive, _) =
+                bound.span().ctxt().outer_expn_data().kind
+            {
+                // We ignore bounds that come from exclusively from a `#[derive(_)]`, because we
```

Comment:
`derive` is irrelevant, the problem is hard to concatenate spans from different syntactic contexts.

If you want this logic, please remove this change from this PR, and add the logic for concatenating a set of potentially incompatible spans to `rustc_span` in a separate PR.
The "concatenatable" subset should come from the same context, but which subset to use if there are multiple such subsets is a question. Using the root one (`SyntaxContext::empty()`) if it exists, would probably be a good start approximation.

Reply:
This change is to avoid the following in the new test:

```
help: consider further restricting this bound
   |
LL |     #[derive(Debug + c::Trait)]
   |                    ++++++++++
```

I understand that we would want to check that all the spans belong to the same context and do something smarter than this, but we know that *this* will happen. `bounds_span` is used in three suggestions (add a type param when `_` used in a field or arg, suggest restricting an existing type param in middle and trait_selection). Because of that, I thought it appropriate to avoid emitting a completely wrong suggestion. I can also modify the caller and avoid calling `bounds_span`, but that seemed worse to me than having this "hidden logic" here. Thoughts?

Reply:
Two questions:
- Is it possible to remove `fn bounds_span` entirely and use a multi-span or something else instead?
It looks like the full combined span isn't even used at the `bounds_span`'s points of use, it's always `shrink_to_hi`'ed before being reported.
- If `bounds_span` is kept, could you move the span-combining logic to a separate span-combining function in `rustc_span`. The `MacroKind::Derive` logic can be kept inside it, but it will need a big FIXME saying that the derive is actually irrelevant and the function needs to be rewritten to be more general. IIRC, `bounds_span` isn't the only place where such span combining is used.

Reply:
- `bounds_span` could return a `MultiSpan` instead, that would make more sense.
 - sounds reasonable.

Reply:
Added `Span::can_be_used_for_suggestions` which checks that it's either on the Root context *or* it is a `derive` macro and the position is different to its parent span. This is only an approximation (that might be brittle), but when we use `self.span` in the compiler's `derive`s the parent span will be the same position. The problem happens is if the `ctxt` depth is higher than one (some macro producing another macro, or whatever a user supplied proc-macro might do), but it *seems* safe enough to me for merging. What do you think?